skalenetwork/skaled
# Support "input" field in eth_call, eth_estimateGas

## Description

This PR adds support of "input" field in `eth_call`, `eth_estimateGas` according to ETH RPC documentation: https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_call. Field "data" will be still supported

## Tests

**Unit test:** 
Enhance `JsonRpcSuite/simple_contract` with new call format

**Manual test:**
1. Run skaled w/o fix (e.g. 3.18.0)
2. Deploy simple contract
```
// SPDX-License-Identifier: GPL-3.0

pragma solidity >=0.7.0 <0.9.0;

contract Storage {

    uint256 a;

    function getA() public view returns (uint256) {
        return a;
    }

    function setA(uint _a) external  {
        a = _a;
    }
}
```
3. Run cmd: 
```
curl ENDPOINT -X POST -H "Content-Type: application/json" -d  '{"id":1931024001,"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"input":"0xd46300fd","from":"0xa68F946090c600eDa6f139783077EE802Afeb990","to":"0x9EcD93c2cF9E551B3B02939c72D3F515A8cb76B0","type":"0x1","value":"0x0"},"latest"]}'
```
4. Run cmd: 
```
curl ENDPOINT -X POST -H "Content-Type: application/json" -d  '{"id":1931024001,"jsonrpc":"2.0","method":"eth_call","params":[{"input":"0xd46300fd","from":"0xa68F946090c600eDa6f139783077EE802Afeb990","to":"0x9EcD93c2cF9E551B3B02939c72D3F515A8cb76B0","type":"0x1","value":"0x0"},"latest"]}'
```
5. Both cmds return `EVM revert instruction without description message`
6. Update skaled, run steps 3-4. Commands should return: 
```
eth_call: {"id":1931024001,"jsonrpc":"2.0","result":"0x000000000000000000000000000000000000000000000000000000000000007b"}
eth_estimateGas: {"id":1931024001,"jsonrpc":"2.0","result":"0x54db"}
```



## Performance Impact

JSON RPC API change only, no performance impact
